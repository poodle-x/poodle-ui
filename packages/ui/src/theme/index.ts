import React from "react";
import {
	ThemeValue,
	SystemThemeConfig,
	BreakpointsThemeConfig,
	SpacingThemeConfig,
	ColorsThemeConfig,
	FontsThemeConfig,
	BordersThemeConfig,
	ShadowThemeConfig,
	LayoutsThemeConfig,
} from "@poodle/system";

import { CSSObjectSystem, Stylesheet } from "../styled";
import { BoxThemeConfig } from "../Box";
import { ButtonThemeConfig } from "../Button";
import { ColumnsThemeConfig } from "../Columns";
import { ColumnThemeConfig } from "../Column";
import { ContainerThemeConfig } from "../Container";
import { TableThemeConfig } from "../Table";
import { TableHeadThemeConfig } from "../TableHead";
import { TableBodyThemeConfig } from "../TableBody";
import { TableRowThemeConfig } from "../TableRow";
import { TableCellThemeConfig } from "../TableCell";
import { VisuallyHiddenThemeConfig } from "../VisuallyHidden";
import { CheckControlThemeConfig } from "../CheckControl";
import { CheckboxThemeConfig } from "../Checkbox";
import { RadioThemeConfig } from "../Radio";
import { InputThemeConfig } from "../Input";
import { InputAdornmentThemeConfig } from "../InputAdornment";
import { IconButtonThemeConfig } from "../IconButton";
import { ModalThemeConfig } from "../Modal";
import { ModalHeaderThemeConfig } from "../ModalHeader";
import { ModalBodyThemeConfig } from "../ModalBody";
import { ModalFooterThemeConfig } from "../ModalFooter";
import { SelectThemeConfig } from "../Select";
import { TabsThemeConfig } from "../Tabs";
import { TabThemeConfig } from "../Tab";
import { TabListThemeConfig } from "../TabList";
import { TabPanelThemeConfig } from "../TabPanel";

export { Stylesheet };

export { useTheme } from "./useTheme";

export type DirectionThemeConfig = "rtl" | "ltr";

export interface GlobalThemeConfig {
	styles?: ThemeValue<CSSObjectSystem>;
}

export type IconType = React.FC<
	React.PropsWithoutRef<React.SVGProps<SVGSVGElement>> &
		React.RefAttributes<SVGSVGElement>
>;

export interface BaseIcons {
	RadioButton?: IconType;
	RadioCheckedButton?: IconType;
	CheckboxButton?: IconType;
	CheckboxCheckedButton?: IconType;
	CheckboxIndeterminateButton?: IconType;
	ArrowDown?: IconType;
}

export interface ThemeConfig extends SystemThemeConfig {
	global?: GlobalThemeConfig;
	direction?: ThemeValue<DirectionThemeConfig>;
	baseIcons?: BaseIcons;
	Box?: BoxThemeConfig;
	Button?: ButtonThemeConfig;
	Columns?: ColumnsThemeConfig;
	Column?: ColumnThemeConfig;
	Container?: ContainerThemeConfig;
	Table?: TableThemeConfig;
	TableHead?: TableHeadThemeConfig;
	TableBody?: TableBodyThemeConfig;
	TableRow?: TableRowThemeConfig;
	TableCell?: TableCellThemeConfig;
	VisuallyHidden?: VisuallyHiddenThemeConfig;
	CheckControl?: CheckControlThemeConfig;
	Checkbox?: CheckboxThemeConfig;
	Radio?: RadioThemeConfig;
	Input?: InputThemeConfig;
	InputAdornment?: InputAdornmentThemeConfig;
	IconButton?: IconButtonThemeConfig;
	Modal?: ModalThemeConfig;
	ModalHeader?: ModalHeaderThemeConfig;
	ModalBody?: ModalBodyThemeConfig;
	ModalFooter?: ModalFooterThemeConfig;
	Select?: SelectThemeConfig;
	Tabs?: TabsThemeConfig;
	Tab?: TabThemeConfig;
	TabList?: TabListThemeConfig;
	TabPanel?: TabPanelThemeConfig;
}

export const defaultGlobal = (
	overrideTheme: ThemeConfig
): GlobalThemeConfig => {
	return {
		styles: overrideTheme.global?.styles,
	};
};

export const defaultBreakpoints = (
	overrideTheme: ThemeConfig
): BreakpointsThemeConfig => {
	return {
		scale: ["sm", "md", "lg", "xl"],
		...overrideTheme.breakpoints,
		sets: {
			sm: 576,
			md: 768,
			lg: 992,
			xl: 1200,
			xxl: 1400,
			...overrideTheme.breakpoints?.sets,
		},
	};
};

export const defaultSpacing = (
	overrideTheme: ThemeConfig
): SpacingThemeConfig => {
	return {
		scale: 4,
		...overrideTheme.spacing,
		sets: {
			s: 4,
			m: 8,
			l: 16,
			...overrideTheme.spacing?.sets,
		},
	};
};

export const defaultFonts = (overrideTheme: ThemeConfig): FontsThemeConfig => {
	return {
		scale: "1rem",
		...overrideTheme.fonts,
		base: {
			fontFamily:
				'-apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol"',
			fontStyle: "normal",
			fontWeight: "normal",
			lineHeight: "125%",
			fontSize: "1rem",
			...overrideTheme.fonts?.base,
		},
		weights: {
			normal: 400,
			semibold: 600,
			bold: 700,
			...overrideTheme.fonts?.weights,
		},
		sets: {
			input: {
				fontSize: "1rem",
				fontWeight: "normal",
				lineHeight: "125%",
			},
			button: {
				fontSize: "1rem",
				fontWeight: "bold",
				lineHeight: "150%",
			},
			label: {
				fontStyle: "normal",
				fontWeight: "bold",
				fontSize: "0.875rem",
				lineHeight: "110%",
			},
			helper: {
				fontStyle: "normal",
				fontWeight: "normal",
				fontSize: "0.875rem",
				lineHeight: "110%",
			},
			checkControlLabelMedium: {
				fontStyle: "normal",
				fontWeight: "normal",
				fontSize: "1rem",
				lineHeight: "110%",
			},
			checkControlLabelSmall: {
				fontStyle: "normal",
				fontWeight: "normal",
				fontSize: "0.875rem",
				lineHeight: "110%",
			},
			checkControlLabelLarge: {
				fontStyle: "normal",
				fontWeight: "normal",
				fontSize: "1.125rem",
				lineHeight: "110%",
			},
			modalTitle: {
				fontStyle: "normal",
				fontWeight: "bold",
				fontSize: "1.5rem",
				lineHeight: "133.33%",
			},
			...overrideTheme.fonts?.sets,
		},
	};
};

export const defaultColor = (overrideTheme: ThemeConfig): ColorsThemeConfig => {
	return {
		modes: {
			...overrideTheme.colors?.modes,
		},
		sets: {
			text: "#282828",
			disabledText: "#A6A6A6",

			primary: {
				light: "#D1EBFF",
				base: "#00478A",
				dark: "#003A70",
				textOnLight: "#00478A",
				textOnBase: "#fff",
				textOnDark: "#fff",
			},

			negative: {
				base: "#AF1308",
				dark: "#8D1007",
				light: "#FCF3F3",
				textOnLight: "#AF1308",
				textOnBase: "#fff",
				textOnDark: "#fff",
			},

			positive: {
				base: "#006122",
				dark: "#004D1B",
				light: "#E5FFEE",
				textOnLight: "#006122",
				textOnBase: "#fff",
				textOnDark: "#fff",
			},

			warn: {
				base: "#FFBB33",
				dark: "#FAAB00",
				light: "#FFF1DB",
				textOnLight: "#282828",
				textOnBase: "#282828",
				textOnDark: "#282828",
			},

			bg: {
				base: "#F8F8F8",
				100: "#E0E0E0",
				textOn100: "#282828",
				200: "#CCCCCC",
				textOn200: "#282828",
			},

			modal: {
				base: "#F8F8F8",
				footer: "#F0F0F0",
			},

			border: {
				base: "#CCCCCC",
			},

			...overrideTheme.colors?.sets,
		},
	};
};

export const defaultBorder = (
	overrideTheme: ThemeConfig
): BordersThemeConfig => {
	return {
		sets: {
			base: {
				style: "solid",
				width: "1px",
				color: "border",
			},
			...overrideTheme?.borders?.sets,
		},
		radius: {
			base: "4px",
			...overrideTheme?.borders?.radius,
		},
	};
};

export const defaultBoxShadow = (
	overrideTheme: ThemeConfig
): ShadowThemeConfig => {
	return {
		boxShadowSets: {
			lv1: "1px 2px 2px -1px rgba(0,0,0,0.4)",
			lv2: "2px 2px 5px 0px rgba(0,0,0,0.4)",
			lv3: "4px 4px 5px 0px rgba(0,0,0,0.4)",
			...overrideTheme?.shadows?.boxShadowSets,
		},
	};
};

export const defaultLayouts = (
	overrideTheme: ThemeConfig
): LayoutsThemeConfig => {
	return {
		zIndices: {
			modal: 2000,
			...overrideTheme?.layouts?.zIndices,
		},
	};
};

export function createTheme(overrideTheme: ThemeConfig = {}): ThemeConfig {
	return {
		direction: "ltr",
		...overrideTheme,
		global: defaultGlobal(overrideTheme),
		spacing: defaultSpacing(overrideTheme),
		breakpoints: defaultBreakpoints(overrideTheme),
		colors: defaultColor(overrideTheme),
		fonts: defaultFonts(overrideTheme),
		borders: defaultBorder(overrideTheme),
		shadows: defaultBoxShadow(overrideTheme),
		layouts: defaultLayouts(overrideTheme),
	};
}

export type StylesConfig<T extends string> = {
	[K in T]: ThemeValue<CSSObjectSystem>;
};

export interface StandardComponentProps {
	variant?: string;
	themeExtend?: ThemeConfig;
	theme?: ThemeConfig;
	sizeStyle?: string;
	colorStyle?: string;
}

export interface StandardThemeConfig<
	T extends string,
	Props extends { [key: string]: any }
> {
	defaultProps?: Partial<Props>;
	styles?: Partial<StylesConfig<T>>;
	variants?: {
		[key: string]: Partial<StylesConfig<T>>;
	};
	sizes?: {
		[key: string]: Partial<StylesConfig<T>>;
	};
	colorStyles?: {
		[key: string]: Partial<StylesConfig<T>>;
	};
	modes?: {
		[key: string]: {
			styles?: Partial<StylesConfig<T>>;
			variants?: {
				[key: string]: Partial<StylesConfig<T>>;
			};
			sizes?: {
				[key: string]: Partial<StylesConfig<T>>;
			};
			colorStyles?: {
				[key: string]: Partial<StylesConfig<T>>;
			};
		};
	};
	overrides?: Partial<StylesConfig<T>>;
	overrideClasses?: Partial<
		{
			[K in T]: boolean | string;
		}
	>;
}

export * from "./system";
