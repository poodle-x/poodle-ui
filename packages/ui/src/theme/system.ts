import {
	BordersProps,
	bordersSystem,
	ColorsProps,
	colorsSystem,
	compose,
	createSystem,
	FlexProps,
	flexSystem,
	FontsProps,
	fontsSystem,
	getMode,
	GridProps,
	gridSystem,
	LayoutProps,
	layoutSystem,
	ListProps,
	listSystem,
	PseudosProps,
	pseudosSystem,
	ResponsiveProp,
	ShadowProps,
	shadowSystem,
	SpacingProps,
	spacingSystem,
	SystemConfigList,
} from "@poodle/system";
import * as CSS from "csstype";
import { CSSSystem } from "../styled";
import { StandardThemeConfig } from "./index";

const additionalSystem: SystemConfigList = {
	objectFit: true,
	objectPosition: true,
	transition: true,
	cursor: true,
	pointerEvents: true,
};

export interface AdditionalSystemProps {
	/**
	 * The CSS `object-fit` property
	 */
	objectFit?: ResponsiveProp<CSS.Property.ObjectFit>;
	/**
	 * The CSS `object-fit` property
	 */
	objectPosition?: ResponsiveProp<CSS.Property.ObjectPosition>;
	/**
	 * The CSS `transition` property
	 */
	transition?: ResponsiveProp<CSS.Property.Transition>;
	/**
	 * The CSS `cursor` property
	 */
	cursor?: ResponsiveProp<CSS.Property.Cursor>;
	/**
	 * The CSS `cursor` property
	 */
	pointerEvents?: ResponsiveProp<CSS.Property.PointerEvents>;
}

const allSystem = compose(
	colorsSystem,
	fontsSystem,
	layoutSystem,
	spacingSystem,
	bordersSystem,
	flexSystem,
	pseudosSystem,
	shadowSystem,
	gridSystem,
	listSystem,
	additionalSystem
);

export const system = createSystem(allSystem);

interface CssSystemProps
	extends SpacingProps,
		LayoutProps,
		FontsProps,
		ColorsProps,
		BordersProps,
		ShadowProps,
		FlexProps,
		GridProps,
		ListProps,
		AdditionalSystemProps {}

export interface SystemProps
	extends SpacingProps,
		LayoutProps,
		FontsProps,
		ColorsProps,
		BordersProps,
		ShadowProps,
		FlexProps,
		PseudosProps<CssSystemProps>,
		GridProps,
		ListProps,
		AdditionalSystemProps {}

const additionalProps = [
	"textStyle",
	"variant",
	"themeExtend",
	"theme",
	"sizeStyle",
	"colorStyle",
];

// Cache to avoid unnecessary repeat loop
// @todo maybe move to a helper func
const finalSystem: SystemConfigList = {};

for (const keys in allSystem) {
	const keysPart = keys.split(",");

	const rawConfig = allSystem[keys];

	keysPart.forEach((k) => {
		finalSystem[k] = rawConfig;
	});
}

export function removeSystemProps(props: { [key: string]: any }) {
	let resultProps = {};

	for (const prop in props) {
		let found = false;

		if (finalSystem[prop] || additionalProps.indexOf(prop) >= 0) {
			found = true;
		}

		if (!found && ["sx"].indexOf(prop) === -1) {
			resultProps = {
				...resultProps,
				[prop]: props[prop],
			};
		}
	}

	return resultProps;
}

export function createCSSSystemStandard<Key extends string>(data: {
	key: Key;
	config?: StandardThemeConfig<Key, any>;
	props: { [key: string]: any };
	base?: CSSSystem["base"];
}): CSSSystem {
	const { key, props, config, base } = data;

	const { variant = "", sizeStyle = "", colorStyle = "" } = props;

	const mode = getMode(props);

	return {
		base,
		applies: [
			config?.styles?.[key],
			config?.modes?.[mode]?.styles?.[key],
			config?.variants?.[variant]?.[key],
			config?.modes?.[mode]?.variants?.[variant]?.[key],
			config?.sizes?.[sizeStyle]?.[key],
			config?.modes?.[mode]?.sizes?.[sizeStyle]?.[key],
			config?.colorStyles?.[colorStyle]?.[key],
			config?.modes?.[mode]?.colorStyles?.[colorStyle]?.[key],
		],
		override: config?.overrides?.[key],
	};
}

export * from "@poodle/system";
