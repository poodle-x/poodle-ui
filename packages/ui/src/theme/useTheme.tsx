import { ThemeContext } from "@emotion/react";
import deepmerge from "deepmerge";
import React from "react";
import { ThemeConfig } from "./";

export function useTheme(
	options: {
		override?: ThemeConfig;
		extend?: ThemeConfig;
	} = {}
): { theme: ThemeConfig } {
	const { override, extend } = options;

	const themeContext: ThemeConfig = React.useContext(ThemeContext);

	const returnTheme = React.useMemo(() => {
		if (override) {
			return override;
		}

		if (extend) {
			return deepmerge(themeContext, extend, {
				arrayMerge: (_, sourceArray) => {
					return sourceArray;
				},
			});
		}

		return themeContext;
	}, [extend, override, themeContext]);

	return {
		theme: returnTheme || {},
	};
}
