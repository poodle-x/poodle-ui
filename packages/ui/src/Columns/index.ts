import { Columns, ColumnsProps } from "./Columns";
import { StandardThemeConfig } from "../theme";

export * from "./Columns";

export default Columns;

export type ColumnsStyleKeys = "root";

export interface ColumnsThemeConfig
	extends StandardThemeConfig<ColumnsStyleKeys, ColumnsProps> {}
