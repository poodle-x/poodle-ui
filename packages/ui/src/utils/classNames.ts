import React from "react";
import { css, CSSObjectSystem, cx } from "../styled";
import { ThemeValue } from "../theme";
import { cssSystem } from "./cssSystem";

type ClassName =
	| undefined
	| string
	| ((props: any) => string)
	| { [key: string]: boolean };

export function genClassNames(data: {
	props: any;
	classNames: Array<ClassName>;
}) {
	const { props, classNames } = data;

	let resultClassNames = "";

	classNames.forEach((cn) => {
		if (typeof cn === "function") {
			resultClassNames = cx(resultClassNames, cn(props));
			return;
		}

		resultClassNames = cx(resultClassNames, cn);
	});

	return resultClassNames;
}

export function classNameCompose(data: {
	props: any;
	overrideClassName?: string | boolean | undefined;
	styles: {
		base: CSSObjectSystem[];
		applies?: (ThemeValue<CSSObjectSystem> | undefined)[];
		override?: ThemeValue<CSSObjectSystem>;
	};
}) {
	const { overrideClassName, styles, props } = data;

	if (typeof overrideClassName === "string") {
		return overrideClassName;
	}

	if (overrideClassName === true) {
		return "";
	}

	return css(
		cssSystem({
			...styles,
			props,
		})
	);
}

export function genListClassNames(
	baseProps: any,
	lists: {
		[key: string]: {
			props?: any;
			classNames: Array<ClassName>;
		};
	}
): { [key in keyof typeof lists]: string } {
	const result: { [key: string]: string } = {};
	Object.keys(lists).forEach((key) => {
		let resultClassNames = "";
		const listKeyData = lists[key];
		resultClassNames = genClassNames({
			props: { ...baseProps, ...listKeyData.props },
			classNames: listKeyData.classNames,
		});
		result[key] = resultClassNames;
	});

	return result;
}

export function useGenListClassNames(
	baseProps: any,
	lists: {
		[key: string]: {
			props?: any;
			classNames: Array<ClassName>;
		};
	}
) {
	return React.useMemo(() => {
		return genListClassNames(baseProps, lists);
	}, [lists, baseProps]);
}
