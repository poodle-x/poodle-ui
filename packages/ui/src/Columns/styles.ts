import {
	getBreakpointsScale,
	ResponsiveProp,
	getSpacing,
	createCSSSystemStandard,
} from "../theme";
import { CSSSystem } from "../styled";
import { ColumnsProps } from "./Columns";
import { responsiveValueThreshold } from "../utils/responsiveValueThreshold";
import { negativeResponsiveValue } from "../utils/negativeResponsiveValue";
import { transformResponsiveValue } from "../utils/transformResponsiveValue";
import { ColumnsStyleKeys } from "./";

export function Root(props: ColumnsProps): CSSSystem {
	const { collapse, gutter, theme, wrapMode, wrapGutter } = props;

	const systemGutter = gutter
		? transformResponsiveValue(gutter, (value) => {
				return getSpacing(value)(props);
		  })
		: gutter;

	const negativeGutter = systemGutter
		? negativeResponsiveValue(systemGutter)
		: systemGutter;

	const systemWrapGutter = wrapGutter
		? transformResponsiveValue(wrapGutter, (value) => {
				return getSpacing(value)(props);
		  })
		: wrapGutter;

	const negativeWrapGutter = systemWrapGutter
		? negativeResponsiveValue(systemWrapGutter)
		: systemWrapGutter;

	const breakPointScale = getBreakpointsScale(props);

	let calcGutter: ResponsiveProp<number | string> | undefined = undefined;

	if (collapse && negativeGutter) {
		calcGutter = responsiveValueThreshold<number | string>(
			negativeGutter,
			collapse,
			breakPointScale
		);
	}
	const wrapModeText = wrapMode ? "wrap" : undefined;

	return createCSSSystemStandard<ColumnsStyleKeys>({
		key: "root",
		config: theme?.Columns,
		props,
		base: [
			{
				display: collapse ? { [collapse]: "flex" } : "flex",
				marginLeft: collapse && calcGutter ? calcGutter : negativeGutter,
				marginTop: negativeWrapGutter,
				flexWrap: collapse && wrapMode ? { [collapse]: "wrap" } : wrapModeText,
			},
		],
	});
}
