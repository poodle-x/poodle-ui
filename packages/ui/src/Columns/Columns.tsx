import React from "react";
import Box, { BoxProps } from "../Box";
import { getCSSSystemBoxProps } from "../styled/CSSSystem";
import { ResponsiveProp, StandardComponentProps, ThemeConfig } from "../theme";
import useDefaultProps from "../hooks/useDefaultProps/useDefaultProps";
import * as styles from "./styles";

export interface ColumnsContextValue {
	wrapMode?: boolean;
	collapse?: string;
	wrapGutter?: ResponsiveProp<number | string>;
	gutter?: ResponsiveProp<number | string>;
}

export const ColumnsContext = React.createContext<ColumnsContextValue>({});

export interface LocalColumnsProps {
	/**
	 * Wrap overflows column
	 */
	wrapMode?: boolean;
	/**
	 * Gutter vertical between each column when collapsed or between each row in wrap mode
	 */
	wrapGutter?: ResponsiveProp<number | string>;
	/**
	 * Force columns wrap below breakpoint.
	 */
	collapse?: string;
	/**
	 * Gutter horizontal between each column when not collapsed
	 */
	gutter?: ResponsiveProp<number | string>;
	children?: React.ReactNode;
}

export interface ColumnsProps
	extends BoxProps,
		StandardComponentProps,
		LocalColumnsProps {}

function getDefaultProps(theme?: ThemeConfig) {
	return theme?.Columns?.defaultProps;
}

export const Columns: React.ForwardRefExoticComponent<
	React.PropsWithoutRef<ColumnsProps> & React.RefAttributes<HTMLElement>
> = React.forwardRef<HTMLElement, ColumnsProps>((_props, ref) => {
	const { props, isLocalTheme } = useDefaultProps<ColumnsProps>(_props, {
		themeDefaultProps: getDefaultProps,
	});

	const {
		children,
		collapse,
		className,
		gutter,
		wrapGutter,
		wrapMode,
		...otherProps
	} = props;

	return (
		<Box
			{...otherProps}
			{...getCSSSystemBoxProps({
				isRoot: true,
				isLocalTheme,
				componentProps: props,
				fnCSSSystem: styles.Root,
				baseClassName: ["poodle-columns"],
			})}
			ref={ref}
		>
			<ColumnsContext.Provider
				value={{
					wrapMode,
					wrapGutter,
					collapse,
					gutter,
				}}
			>
				{children}
			</ColumnsContext.Provider>
		</Box>
	);
});

Columns.displayName = "PoodleColumns";
