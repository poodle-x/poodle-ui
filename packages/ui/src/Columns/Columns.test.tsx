import React from "react";
import { render } from "../utils/test";
import Columns, { ColumnsContext } from "./";
import { ThemeConfig } from "../theme";

describe("<Columns />", function () {
	it("should render columns correctly", () => {
		const { container } = render(<Columns>Hello</Columns>);
		expect(container.firstChild).toMatchSnapshot();
	});

	it("ref should work", () => {
		const ref = React.createRef<HTMLDivElement>();

		render(<Columns ref={ref}>Columns</Columns>);

		expect(ref.current).toMatchSnapshot();
	});

	it("should render Box props correctly", () => {
		const { container } = render(
			<Columns aria-disabled="true" id="columns" as="section" p={10}>
				Columns
			</Columns>
		);
		expect(container.firstChild).toMatchSnapshot();
	});

	it("should render with wrap", () => {
		const { container } = render(
			<Columns wrapMode={true} gutter="50px">
				Columns
			</Columns>
		);
		expect(container.firstChild).toMatchSnapshot();
	});

	it("should render with custom gutter correctly", () => {
		const { container } = render(
			<Columns wrapGutter="30px" gutter="50px">
				Columns
			</Columns>
		);
		expect(container.firstChild).toMatchSnapshot();
	});

	it("should render with array responsive gutter correctly", () => {
		const { container } = render(
			<Columns
				wrapGutter={["scale-4", 25, 35, 45]}
				gutter={["scale-4", 20, 30, 40]}
			>
				Columns
			</Columns>
		);
		expect(container.firstChild).toMatchSnapshot();
	});

	it("should render with object responsive gutter correctly", () => {
		const { container } = render(
			<Columns
				wrapGutter={{
					sm: 15,
					md: 25,
					lg: 35,
					xl: 45,
				}}
				gutter={{
					_: 5,
					sm: 10,
					md: 20,
					lg: 30,
					xl: 40,
				}}
			>
				Columns
			</Columns>
		);
		expect(container.firstChild).toMatchSnapshot();
	});

	it("should render with custom collapse correctly", () => {
		const { container } = render(<Columns collapse="md">Columns</Columns>);
		expect(container.firstChild).toMatchSnapshot();
	});

	it("should render with custom collapse and gutter correctly", () => {
		const { container } = render(
			<Columns gutter="20px" collapse="md">
				Columns
			</Columns>
		);
		expect(container.firstChild).toMatchSnapshot();
	});

	it("should render with custom collapse and responsive gutter correctly", () => {
		const { container } = render(
			<Columns
				gutter={{
					_: 5,
					sm: 10,
					md: 20,
					lg: 30,
					xl: 40,
				}}
				collapse="md"
			>
				Columns
			</Columns>
		);
		expect(container.firstChild).toMatchSnapshot();
	});

	it("should send context to children correctly", () => {
		const TestComp = function () {
			const context = React.useContext(ColumnsContext);

			return (
				<div data-testid="test">
					{JSON.stringify(context.gutter)}|{JSON.stringify(context.wrapGutter)}|
					{JSON.stringify(context.collapse)}|{JSON.stringify(context.wrapMode)}
				</div>
			);
		};

		const gutter = {
			xs: 10,
			sm: 20,
			md: 30,
			lg: 40,
		};

		const wrapGutter = {
			xs: 15,
			sm: 25,
			md: 35,
			lg: 45,
		};

		const { getByTestId } = render(
			<Columns
				gutter={gutter}
				wrapGutter={wrapGutter}
				wrapMode={true}
				collapse="md"
			>
				<TestComp />
			</Columns>
		);

		expect(getByTestId("test").innerHTML).toEqual(
			`${JSON.stringify(gutter)}|${JSON.stringify(wrapGutter)}|"md"|true`
		);
	});

	describe("theming", () => {
		const themeTest: ThemeConfig = {
			mode: "mode",
			Columns: {
				styles: {
					root: {
						mt: 10,
						top: 0,
						textIndent: 1,
					},
				},
				variants: {
					custom: {
						root: {
							letterSpacing: 200,
						},
					},
				},
			},
		};

		it("should extend styles in global", () => {
			const { container } = render(
				<Columns variant="custom">Columns</Columns>,
				{
					theme: themeTest,
				}
			);
			expect(container.firstChild).toMatchSnapshot();
		});

		it("should extend styles in local", () => {
			const { container } = render(
				<Columns variant="custom" themeExtend={themeTest}>
					Columns
				</Columns>
			);
			expect(container.firstChild).toMatchSnapshot();
		});

		it("should render with default props correctly", () => {
			const { container } = render(<Columns>Columns</Columns>, {
				theme: {
					Columns: {
						defaultProps: {
							gutter: 50,
							id: "id",
							className: "hello",
						},
					},
				},
			});
			expect(container.firstChild).toMatchSnapshot();
		});

		it("should disable classes", () => {
			const { container } = render(<Columns>Columns</Columns>, {
				theme: {
					Columns: {
						overrideClasses: {
							root: true,
						},
					},
				},
			});
			expect(container.firstChild).toMatchSnapshot();
		});

		it("should override classes", () => {
			const { container } = render(<Columns>Columns</Columns>, {
				theme: {
					Columns: {
						overrideClasses: {
							root: "morning",
						},
					},
				},
			});
			expect(container.firstChild).toMatchSnapshot();
		});

		it("should override styles", () => {
			const { container } = render(<Columns>Columns</Columns>, {
				theme: {
					Columns: {
						overrides: {
							root: {
								mt: 50,
							},
						},
					},
				},
			});
			expect(container.firstChild).toMatchSnapshot();
		});
	});
});
