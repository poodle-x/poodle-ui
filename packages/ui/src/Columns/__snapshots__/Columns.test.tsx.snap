// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Columns /> ref should work 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

<div
  class="poodle-box poodle-columns c0"
>
  Columns
</div>
`;

exports[`<Columns /> should render Box props correctly 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 10px;
}

<section
  aria-disabled="true"
  class="poodle-box poodle-columns c0"
  id="columns"
>
  Columns
</section>
`;

exports[`<Columns /> should render columns correctly 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

<div
  class="poodle-box poodle-columns c0"
>
  Hello
</div>
`;

exports[`<Columns /> should render with array responsive gutter correctly 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-left: -16px;
  margin-top: -16px;
}

@media screen and (min-width: 576px) {
  .c0 {
    margin-left: -20px;
  }
}

@media screen and (min-width: 768px) {
  .c0 {
    margin-left: -30px;
  }
}

@media screen and (min-width: 992px) {
  .c0 {
    margin-left: -40px;
  }
}

@media screen and (min-width: 576px) {
  .c0 {
    margin-top: -25px;
  }
}

@media screen and (min-width: 768px) {
  .c0 {
    margin-top: -35px;
  }
}

@media screen and (min-width: 992px) {
  .c0 {
    margin-top: -45px;
  }
}

<div
  class="poodle-box poodle-columns c0"
>
  Columns
</div>
`;

exports[`<Columns /> should render with custom collapse and gutter correctly 1`] = `
@media screen and (min-width: 768px) {
  .c0 {
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
  }
}

@media screen and (min-width: 768px) {
  .c0 {
    margin-left: -20px;
  }
}

<div
  class="poodle-box poodle-columns c0"
>
  Columns
</div>
`;

exports[`<Columns /> should render with custom collapse and responsive gutter correctly 1`] = `
@media screen and (min-width: 768px) {
  .c0 {
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
  }
}

@media screen and (min-width: 768px) {
  .c0 {
    margin-left: -20px;
  }
}

@media screen and (min-width: 992px) {
  .c0 {
    margin-left: -30px;
  }
}

@media screen and (min-width: 1200px) {
  .c0 {
    margin-left: -40px;
  }
}

<div
  class="poodle-box poodle-columns c0"
>
  Columns
</div>
`;

exports[`<Columns /> should render with custom collapse correctly 1`] = `
@media screen and (min-width: 768px) {
  .c0 {
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
  }
}

<div
  class="poodle-box poodle-columns c0"
>
  Columns
</div>
`;

exports[`<Columns /> should render with custom gutter correctly 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-left: -50px;
  margin-top: -30px;
}

<div
  class="poodle-box poodle-columns c0"
>
  Columns
</div>
`;

exports[`<Columns /> should render with object responsive gutter correctly 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-left: -5px;
}

@media screen and (min-width: 576px) {
  .c0 {
    margin-left: -10px;
  }
}

@media screen and (min-width: 768px) {
  .c0 {
    margin-left: -20px;
  }
}

@media screen and (min-width: 992px) {
  .c0 {
    margin-left: -30px;
  }
}

@media screen and (min-width: 1200px) {
  .c0 {
    margin-left: -40px;
  }
}

@media screen and (min-width: 576px) {
  .c0 {
    margin-top: -15px;
  }
}

@media screen and (min-width: 768px) {
  .c0 {
    margin-top: -25px;
  }
}

@media screen and (min-width: 992px) {
  .c0 {
    margin-top: -35px;
  }
}

@media screen and (min-width: 1200px) {
  .c0 {
    margin-top: -45px;
  }
}

<div
  class="poodle-box poodle-columns c0"
>
  Columns
</div>
`;

exports[`<Columns /> should render with wrap 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-left: -50px;
  -webkit-box-flex-wrap: wrap;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

<div
  class="poodle-box poodle-columns c0"
>
  Columns
</div>
`;

exports[`<Columns /> theming should extend styles in global 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-top: 10px;
  top: 0;
  text-indent: 1px;
  letter-spacing: 200px;
}

<div
  class="poodle-box poodle-columns c0"
>
  Columns
</div>
`;

exports[`<Columns /> theming should extend styles in local 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-top: 10px;
  top: 0;
  text-indent: 1px;
  letter-spacing: 200px;
}

<div
  class="poodle-box poodle-columns c0"
>
  Columns
</div>
`;

exports[`<Columns /> theming should override styles 1`] = `
.c0 {
  margin-top: 50px;
}

<div
  class="poodle-box poodle-columns c0"
>
  Columns
</div>
`;

exports[`<Columns /> theming should render with default props correctly 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-left: -50px;
}

<div
  class="poodle-box poodle-columns hello c0"
  id="id"
>
  Columns
</div>
`;
